File: 11. Trees/110. Balanced Binary Tree/Solution.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.post_order(root) != -1
    
    def post_order(self, root):
        if not root:
            return 0
        
        left = self.recurse(root.left)
        right = self.recurse(root.right)

        if abs(left - right) > 1:
            return -1

        if left == -1 or right == -1:
            return -1
        
        return max(left, right) + 1

File: 11. Trees/110. Balanced Binary Tree/Solution.cpp
#include <cmath>
#include <algorithm>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return recurse(root) != -1;
    }

    int recurse(TreeNode* root)
    {
        if (root == nullptr)
        {
            return 0;
        }

        int left = recurse(root->left);
        int right = recurse(root->right);

        if (std::abs(left - right) > 1)
        {
            return -1;
        }

        if (left == -1 || right == -1)
        {
            return -1;
        }

        return std::max(left, right) + 1;
    }
};